src/anonymousUserTracking/criteriaCompletionChecker.ts

```diff
@@ -0,0 +1,554 @@
+import {
+  SHARED_PREFS_EVENT_TYPE,
+  KEY_ITEMS,
+  TRACK_PURCHASE,
+  TRACK_UPDATE_CART,
+  TRACK_EVENT,
+  UPDATE_CART,
+  UPDATE_USER,
+  KEY_EVENT_NAME,
+  SHARED_PREFS_EVENT_LIST_KEY,
+  SHARED_PREF_MATCHED_CRITERIAS
+} from '../constants';
+
+interface SearchQuery {
+  combinator: string;
+  searchQueries: SearchQuery[] | Criteria[];
+  dataType?: string;
+  searchCombo?: SearchQuery;
+  field?: string;
+  comparatorType?: string;
+  value?: string;
+  fieldType?: string;
+  minMatch?: number;
+  maxMatch?: number;
+}
+
+interface Criteria {
+  criteriaId: string;
+  name: string;
+  createdAt: number;
+  updatedAt: number;
+  searchQuery: SearchQuery;
+}
+
+class CriteriaCompletionChecker {
+  private localStoredEventList: any[];
+
+  constructor(localStoredEventList: string) {
+    this.localStoredEventList = JSON.parse(localStoredEventList);
+  }
+
+  public getMatchedCriteria(criteriaData: string): string | null {
+    let criteriaId: string | null = null;
+
+    try {
+      const json = JSON.parse(criteriaData);
+      if (json.criterias) {
+        criteriaId = this.findMatchedCriteria(json.criterias);
+      }
+    } catch (e) {
+      this.handleJSONException(e);
+    }
+
+    return criteriaId;
+  }
+
+  private findMatchedCriteria(criteriaList: Criteria[]): string | null {
+    let criteriaId: string | null = null;
+    const criteriaIdList = criteriaList.map((criteria) => criteria.criteriaId);
+    const eventsToProcess = this.prepareEventsToProcess(criteriaIdList);
+    criteriaList.forEach((criteria) => {
+      if (!criteriaId && criteria.searchQuery && criteria.criteriaId) {
```

Comment:
If the goal here is to just get the first result that has `.searchQuery` and `.criteriaId` where `result` is defined, we should just do an `array.find` where all of those conditions are met instead. Then you don't have to worry about breaking out of the foreach at all.

Reply:
@mprew97 we will try to improve this. Thanks.